src/ripple/rpc/handlers/WalletPropose.cpp

```diff
@@ -35,14 +38,58 @@ Json::Value walletPropose (Json::Value const& params)
     RippleAddress   naSeed;
     RippleAddress   naAccount;
 
-    if (!params.isMember (jss::passphrase))
-        naSeed.setSeedRandom ();
+    KeyType type = KeyType::secp256k1;
 
-    else if (!naSeed.setSeedGeneric (params[jss::passphrase].asString ()))
+    bool const has_key_type   = params.isMember (jss::key_type);
+    bool const has_passphrase = params.isMember (jss::passphrase);
+
+    if (has_key_type)
+    {
+        // `key_type` must be valid if present.
+
+        type = keyTypeFromString (params[jss::key_type].asString());
+
+        if (isInvalid (type))
+        {
+            return rpcError (rpcBAD_SEED);
+        }
+
+        naSeed = getSeedFromRPC (params);
+    }
+    else if (has_passphrase)
+    {
+        naSeed.setSeedGeneric (params[jss::passphrase].asString());
+    }
+    else
+    {
+        naSeed.setSeedRandom();
+    }
+
+    if (!naSeed.isSet())
+    {
         return rpcError(rpcBAD_SEED);
+    }
+
+    if (type == KeyType::secp256k1)
+    {
+        RippleAddress naGenerator = RippleAddress::createGeneratorPublic (naSeed);
+        naAccount.setAccountPublic (naGenerator, 0);
+    }
+    else if (type == KeyType::ed25519)
+    {
+        uint256 secretkey = keyFromSeed (naSeed.getSeed());
+
+        Blob publickey (33);
+        publickey[0] = 0xED;
+        ed25519_publickey (secretkey.data(), &publickey[1]);
+        secretkey.zero();  // security erase
 
-    RippleAddress naGenerator = RippleAddress::createGeneratorPublic (naSeed);
-    naAccount.setAccountPublic (naGenerator, 0);
+        naAccount.setAccountPublic (publickey);
+    }
+    else
+    {
+        assert (!!! "not reached");
```

Comment:
Thanks.

On Fri, Mar 6, 2015 at 5:54 PM, Josh Juran <notifications@github.com> wrote:

> In src/ripple/rpc/handlers/WalletPropose.cpp
> <https://github.com/ripple/rippled/pull/907#discussion_r25985435>:
>
> >
> > -    RippleAddress naGenerator = RippleAddress::createGeneratorPublic (naSeed);
> > -    naAccount.setAccountPublic (naGenerator, 0);
> > +        naAccount.setAccountPublic (publickey);
> > +    }
> > +    else
> > +    {
> > +        assert (!!! "not reached");
>
> That's because I only changed the other one. Now fixed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ripple/rippled/pull/907/files#r25985435>.
>



-- 
*Edward Hennis | *Senior Software Engineer | *Ripple Labs Inc.*
ed@ripple.com | www.ripple.com | *www.ripplelabs.com
<http://www.ripplelabs.com/>*
T. 352.327.8769